app/services/globus_digest_generator.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+# GlobusDigestGenerator
+#
+# Add digests to Globus DROs if the files currently lack them. Existing files
+# will not be overwritten.
+class GlobusDigestGenerator
+  # @param [Cocina::Model] cocina a DRO to add digests to
+  # @param [Hash] globus_ids a mapping of filenames to their location on disk
+  # @param [Boolean] validate whether to build the new DRO using validation
+  # @return [Cocina::Model] a new Cocina object with (potentially) new digests
+  def self.generate(cocina:, globus_ids:, validate: true)
+    new(cocina: cocina, globus_ids: globus_ids, validate: validate).generate
+  end
+
+  def initialize(cocina:, globus_ids:, validate:)
+    @cocina = cocina
+    @globus_ids = globus_ids
+    @validate = validate
+  end
+
+  def generate
+    return cocina unless cocina.dro?
+
+    props = cocina.to_h
+    props[:structural] = gen_filesets(props[:structural])
+
+    if validate
+      Cocina::Models.build(props)
+    else
+      Cocina::Models.build_request(props)
+    end
+  end
+
+  private
+
+  attr_reader :cocina, :globus_ids, :validate
+
+  def gen_filesets(filesets)
+    filesets[:contains] = filesets[:contains].map do |fileset|
+      gen_files(fileset)
+    end
+
+    filesets
+  end
```

Comment:
If the intent here is to take a structural (an array of file sets), map a function over every file set, and then return the result, we could also consider something like:

```suggestion
  def gen_filesets(structural)
    structural.tap do |file_sets|
      file_sets[:contains].map!(&method(:gen_files))
    end
  end
```

Less readable? 🤔

Reply:
I actually started with a `tap` but then couldn't quite get `generate_filesets` to work using a similar approach. I ended up thinking there was value in keeping the different generate methods structured similarly. I'd be happy to revisit though.

Reply:
OK, np.